### PR TITLE
[TIR] Fix Ramp int32~64 mismatch in VectorizeLoop and NarrowDataType passes

### PR DESCRIPTION
**Pull request by lazycal**
Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

The following model
```python
import tvm
from tvm import relay
import numpy as np

xshape = (1, 1, 1)
inp = np.random.uniform(size=xshape).astype(np.int64)

x = relay.var("x", shape=xshape, dtype='int64')
x = relay.cast(x, 'int64')
x = relay.broadcast_to(x, relay.const([1, 2, 2], dtype='int64'))
func = relay.Function(relay.analysis.free_vars(x), -x)
mod = tvm.IRModule.from_expr(func)

with tvm.transform.PassContext(opt_level=0):
    relay.create_executor("debug", mod, tvm.cpu()).evaluate()(inp)
```
triggers two issues regarding `base` and `stride` dtype mismatch in `Ramp`, one in VectorizeLoop Pass and the other in NarrowDataType Pass. Error message looks like `Check failed: stride.dtype() == base.dtype() (int32 vs. int64) :`. 

### The fix
- During VectorizeLoop, a loop variable will be converted to a ramp but always of dtype `int32`. This PR changes it to use the loop variable's dtype. 
- During NarrowDataType, it can happen that the `stride `is inferred with `int32`, but `base` is not (see the added test case for detail). This PR adds an upcasting when rewriting a `Ramp` node that has `base` and `stride` inferred with different number of bits.

### CODE REVIEW

**Comment by lazycal** [comment]
Not sure whom I should request for reviews, but it seems simimlar to this PR #9582. So ccing the reviewers there @YuchenJin @junrushao1994 @Mousius

**Comment by tqchen** [comment]
also cc @vinx13 @hzfan @yzhliu pelease take a look when you have time

File: src/tir/transforms/narrow_datatype.cc [hunk 1]
```diff
@@ -253,6 +253,23 @@ class DataTypeRewriter : public StmtExprMutator {
...
   }
 
+  PrimExpr VisitExpr_(const RampNode* op) final {
+    PrimExpr base = VisitExpr(op->base);
+    PrimExpr stride = VisitExpr(op->stride);
+    if (base.same_as(op->base) && stride.same_as(op->stride)) {
+      return GetRef<PrimExpr>(op);
+    } else {
+      if (base.dtype().is_int()) {
+        ICHECK(stride.dtype().is_int()) << "Ramp base is int but stride is " << stride.dtype();
```
Comment (ganler) [inline src/tir/transforms/narrow_datatype.cc:263]: I think we can simply assume that base and stride should be of integer types. However, I also noticed that in 

https://github.com/apache/tvm/blob/22c488e3a829ad700de6547be6096fb2d1f02e81/src/tir/ir/expr.cc#L705

Such assumptions are not checked. I am a bit curious if there will be, say `base/stride` in float types?

**Review by ganler — APPROVED** [review]
LGTM and thanks for the contribution! But I am wondering if we can simply perform compatibility casting when constructing Ramp node? (https://github.com/apache/tvm/tree/main/src/tir/ir/expr.cc#L705)

We should add `ICHECK(base.is_int() && stride.is_int())` if the two will only be integers.

**Comment by ganler** [comment]
@lazycal  Using this impl in fixes the bug you mentioned and might be more general to overcome other hidden ones if we can assume that `base` and `stride` must be of integers.

```c++
Ramp::Ramp(PrimExpr base, PrimExpr stride, int lanes, Span span) {
  ICHECK(base.defined());
  ICHECK(stride.defined());
  ICHECK(base.dtype().is_scalar());
  ICHECK(stride.dtype().is_scalar());
  ICHECK_GT(lanes, 1);
  ICHECK(base.dtype().is_int());
  ICHECK(stride.dtype().is_int());
  
  if (base.dtype() != stride.dtype()) {
    size_t bits = std::max(base.dtype().bits(), stride.dtype().bits());
    DataType dtype = base.dtype().with_bits(bits);
    if (base.dtype() != dtype) base = cast(dtype, base);
    if (stride.dtype() != dtype) stride = cast(dtype, stride);
  }

  ObjectPtr<RampNode> node = make_object<RampNode>();
  node->dtype = base.dtype().with_lanes(lanes);
  node->base = base;
  node->stride = stride;
  node->lanes = lanes;
  node->span = std::move(span);
  data_ = std::move(node);
}
```

**Comment by junrushao1994** [comment]
This is definitely an interesting bug! CC: @vinx13 @yzhliu @hzfan would be great if you guys could take a look :-)

**Comment by ganler** [comment]
@lazycal Fair consideration! 

I also tried `./tests/scripts/task_python_unittest.sh` for the direct fix. It seems to pass all related tests (except a few ones due to my environment), which means at least the unit-tests use `Ramp` with `base` and `stride` in `int` and is good with casting for `int64` and `int32` during Ramp node construction.

**Comment by masahi** [comment]
I've just hit a similar error, when compiling an int8 model with tensorized ops (VNNI):

```
  21: tvm::te::MakeTensorize(tvm::te::ComputeOpNode const*, tvm::te::Stage const&, std::unordered_map<tvm::tir::IterVar, tvm::Range, std::hash<tvm::tir::IterVar>, std::equal_to<tvm::tir::IterVar>, std::allocator<std::pair<tvm::tir::IterVar const, tvm::Range> > > const&, bool)
  20: tvm::te::VerifyTensorizeBody(tvm::te::ComputeOpNode const*, tvm::te::Stage const&, std::unordered_map<tvm::tir::IterVar, tvm::PrimExpr, std::hash<tvm::tir::IterVar>, std::equal_to<tvm::tir::IterVar>, std::allocator<std::pair<tvm::tir::IterVar const, tvm::PrimExpr> > > const&, std::unordered_map<tvm::tir::IterVar, tvm::Range, std::hash<tvm::tir::IterVar>, std::equal_to<tvm::tir::IterVar>, std::allocator<std::pair<tvm::tir::IterVar const, tvm::Range> > > const&, std::unordered_map<tvm::tir::IterVar, tvm::Range, std::hash<tvm::tir::IterVar>, std::equal_to<tvm::tir::IterVar>, std::allocator<std::pair<tvm::tir::IterVar const, tvm::Range> > > const&, std::unordered_map<tvm::te::Tensor, tvm::runtime::Array<tvm::Range, void>, std::hash<tvm::te::Tensor>, std::equal_to<tvm::te::Tensor>, std::allocator<std::pair<tvm::te::Tensor const, tvm::runtime::Array<tvm::Range, void> > > > const&, tvm::te::TensorIntrin const&)
  19: tvm::te::MatchTensorizeBody(tvm::te::ComputeOpNode const*, tvm::te::Stage const&, std::unordered_map<tvm::tir::IterVar, tvm::Range, std::hash<tvm::tir::IterVar>, std::equal_to<tvm::tir::IterVar>, std::allocator<std::pair<tvm::tir::IterVar const, tvm::Range> > > const&, std::unordered_map<tvm::tir::IterVar, tvm::Range, std::hash<tvm::tir::IterVar>, std::equal_to<tvm::tir::IterVar>, std::allocator<std::pair<tvm::tir::IterVar const, tvm::Range> > > const&, std::unordered_map<tvm::te::Tensor, tvm::runtime::Array<tvm::Range, void>, std::hash<tvm::te::Tensor>, std::equal_to<tvm::te::Tensor>, std::allocator<std::pair<tvm::te::Tensor const, tvm::runtime::Array<tvm::Range, void> > > > const&, tvm::te::TensorIntrin const&, tvm::runtime::Map<tvm::tir::Var, tvm::Range, void, void>*)
  18: non-virtual thunk to tvm::tir::StmtExprMutator::VisitExpr(tvm::PrimExpr const&)
  17: _ZZN3tvm3tir11ExprFunctorIFNS_8PrimExprERKS2_EE10InitVTableEvENUlRKNS_7runtime
  16: tvm::te::TensorIntrinMatcher::VisitExpr_(tvm::tir::ReduceNode const*)

  ...

  0: tvm::tir::FloorDiv::FloorDiv(tvm::PrimExpr, tvm::PrimExpr, tvm::Span)
  File "/home/masa/projects/dev/tvm/src/tir/ir/expr.cc", line 322
TypeError: Check failed: (a.dtype() == b.dtype()) is false: mismatched types. int64 vs. int32
```

I wonder if this is related.

**Review by masahi — APPROVED** [review]

**Pull request merged**